src/components/CheckEditor/CheckEditor.tsx

```diff
@@ -12,207 +12,79 @@ import {
-  };
-
-  onSettingsUpdate = ({ settings, labels }: OnUpdateSettingsArgs) => {
-    this.setState(state => {
-      const check = state.check as Check;
-      check.settings = settings;
-      check.labels = labels ?? [];
-      return { check };
-    });
-  };
-
-  onSetType = (type: SelectableValue<CheckType>) => {
-    let check = { ...this.state.check } as Check;
-    if (!type.value) {
-      return;
-    }
-    const typeOfCheck = type.value;
-    const settings = defaultSettings(typeOfCheck);
-    if (!settings) {
-      return;
-    }
-    check.settings = settings;
-
-    this.setState({ check, typeOfCheck });
-  };
-
-  onJobUpdate = (event: React.ChangeEvent<HTMLInputElement>) => {
-    let check = { ...this.state.check } as Check;
-    check.job = event.target.value;
-    this.setState({ check });
-  };
-
-  onTargetUpdate = (target: string) => {
-    let check = { ...this.state.check } as Check;
-    check.target = target;
-    this.setState({ check });
-  };
+export const CheckEditor: FC<Props> = ({ check, instance, onReturn }) => {
+  const [showDeleteModal, setShowDeleteModal] = useState(false);
+  const [error, setError] = useState<SubmissionError | undefined>();
 
-  onProbeOptionsChange = ({ timeout, frequency, probes }: OnChangeArgs) => {
-    this.setState(state => ({
-      check: {
-        ...state.check,
-        timeout,
-        frequency,
-        probes,
-      },
-    }));
-  };
+  const defaultValues = useMemo(() => getDefaultValuesFromCheck(check), [check]);
 
-  onFrequencyUpdate = (event: React.ChangeEvent<HTMLInputElement>) => {
-    let check = { ...this.state.check } as Check;
-    check.frequency = event.target.valueAsNumber * 1000;
-    this.setState({ check });
-  };
+  const formMethods = useForm<CheckFormValues>({ defaultValues, mode: 'onChange' });
+  const selectedCheckType = formMethods.watch('checkType').value as CheckType;
 
-  onTimeoutUpdate = (event: React.ChangeEvent<HTMLInputElement>) => {
-    let check = { ...this.state.check } as Check;
-    check.timeout = event.target.valueAsNumber * 1000;
-    this.setState({ check });
-  };
+  let isEditor = hasRole(OrgRole.EDITOR);
 
-  onSave = async () => {
-    const { instance } = this.props;
-    const { check } = this.state;
+  const onSubmit = async (values: CheckFormValues) => {
+    const updatedCheck = getCheckFromFormValues(values, defaultValues);
     try {
       if (check.id) {
-        await instance.updateCheck(check);
+        await instance.updateCheck({
+          id: check.id,
+          tenantId: check.tenantId,
+          ...updatedCheck,
+        });
       } else {
-        await instance.addCheck(check);
+        await instance.addCheck(updatedCheck);
       }
-      this.props.onReturn(true);
+      onReturn(true);
     } catch (e) {
-      this.setState({
-        error: {
-          status: e.status,
-          message: e.data?.message ?? 'Something went wrong',
-        },
-      });
+      setError(e);
```

Comment:
We've discussed this with @s-h-a-d-o-w  a bit,
using hooks after `await` in a component has potential problems:
- possible race condition if called multiple times
- if component is umounted in the meanwhile, hooks like `setError` will throw

Consider using https://github.com/slorber/react-async-hook  (`useAsyncCabllack` specifically) or similar for things like `onSubmit`?

Reply:
Thanks for that, good to know!